### PR TITLE
HP-1778: update keycloak login

### PR DESCRIPTION
**Pull request by NikoHelle**
Keycloak (Helsinki-Tunnistus) can now be used for logging in other ways than just plain Suomi.fi

Removed "mvp" and "plainSuomiFi" references and PLAIN_SUOMIFI env vars.

The biggest changes are in api tokens, because Keycloak returns them one by one and Tunnistamo returns all tokens defined in the scope. Keycloak has also extra params when fetching tokens, which are not used with Tunnistamo

### CODE REVIEW

**Review by akikoskinen — CHANGES_REQUESTED** [review]
I was somewhat confused checking the first ten or so commits, because they seem to build something related to plain suomi.fi and that was exactly the thing that was supposed to be removed. Then a commit came that renamed everything "plain suomi.fi" to "keycloak". Quite confusing to develop a feature with the wrong name and then rename in the end. I suggest using the correct terminology right from the beginning.

I also tested this. The Tunnistamo side seemed to work nicely. But the Helsinki-Tunnistus side gets logged in, and then after about a second, automatically logged out. I don't quite know what's going on there. Perhaps something related to "silent renew" as the browser makes a request to `http://localhost:3000/silent_renew.html?error=login_required&state=847a7ca54cd944f5b892f32a24d2f253`.

**Comment by NikoHelle** [comment]
Main reason for using old terminology in first commits, was that the docs had conflicting naming and did not want wait for final decision. And maybe end up renaming twice.

Also this is a quite big change. I had to test first, what needs to be changed to make it work with both OIDC servers. Before making bigger changes by renaming a lot of stuff, I was better to make small changes one by one and then verify everything still works. 

I've had Example UI open with Helsinki-Tunnistus login now for 40 minutes without errors. Session is renewing. Silent_renew should not even trigger until session is ending (in 300 secs) and that won't happen in one second.  Maybe a cache problem? It is very sticky with Profile and Example.

**Comment by akikoskinen** [comment]
> I've had Example UI open with Helsinki-Tunnistus login now for 40 minutes without errors. Session is renewing. Silent_renew should not even trigger until session is ending (in 300 secs) and that won't happen in one second. Maybe a cache problem? It is very sticky with Profile and Example.

The issue appeared on Firefox. I tried invalidating any caches and using a private window but the problem persists. I tried Chromium and there it seems to work just fine. Some small difference between browsers?

Firefox prints to console some warnings about cookies like this:

    Cookie “sessionid” with the “SameSite” attribute value “Lax” or “Strict” was omitted because of a cross-site redirect.

I'm not sure what "omitted" means exactly, but if the browser just discards some cookies that could perhaps cause some problems?

**Comment by akikoskinen** [comment]
> Main reason for using old terminology in first commits, was that the docs had conflicting naming and did not want wait for final decision. And maybe end up renaming twice.
> 
> Also this is a quite big change. I had to test first, what needs to be changed to make it work with both OIDC servers. Before making bigger changes by renaming a lot of stuff, I was better to make small changes one by one and then verify everything still works.

Things change and plans change, that's usual. Commit history can be reworked until it's merged. Now there's an end result. The history how that end result is achieved, can be rewritten and polished. Future generations (and reviewers) will praise the author.

**Comment by NikoHelle** [comment]
Firefox (nor Safari) do not work locally with the version in the master branch either. Reason is third party cookies used by Keycloak. Don't know how Tunnistamo differs, as it works like a charm. Problem does not occur when example-ui.*.hel.ninja is used with tunnistus.*.hel.ninja, because then cookies are in same domain and not considered as third party.

I added info about this to the README.md and also how to test locally by changing browser's security settings.

**Comment by NikoHelle** [comment]
I still think it was better to make small adjustments first, before changing/renaming all keycloak settings. It's easier to understand and also if, for some odd reason, only the audience changes were cherry-picked etc, then those are easy to deploy, because all vars in the Azure Library does not have to be renamed.  All commits should work alone and be deployable.

I added  two env vars in the first commit:
REACT_APP_PLAIN_SUOMIFI_EXAMPLE_API_TOKEN_AUDIENCE
REACT_APP_PLAIN_SUOMIFI_PROFILE_API_TOKEN_AUDIENCE

Because Keycloak requires an audience for each api token.  Next commits use these audiences. Later on when everything works and tests succeed, env vars are renamed.

Reversing these changes does not make this PR more clearer or beneficial, imho.

File: src/client/index.ts
Comment (akikoskinen) [inline src/client/index.ts:15]: Perhaps the `TokenPayload` type doesn't need to be exported outside of this client module. I tried to look in all places (excluding tests) where it's used and it seems that it's actually always `JWTPayload`. There is some code that seems to handle also `KeyCloakApiTokenPayload`, but does that code ever get executed in reality?

File: README.md
Comment (akikoskinen) [inline README.md:26]: There's an unpaired quotation mark.

**Review by akikoskinen — COMMENTED** [review]
Some comments. I didn't yet go through the whole PR.

Commit messages have three different ways to mark the JIRA ticket number. Perhaps one style is enough.

File: src/apiAccessTokens/useApiAccessTokens.ts
Comment (NikoHelle) [reply src/apiAccessTokens/useApiAccessTokens.ts]: It works also that way, but true, that is unneccessarily complex comparison. Changed.

File: src/client/index.ts
Comment (NikoHelle) [reply src/client/index.ts]: Yeah. Changed also other plurals to singular.

Comment (NikoHelle) [reply src/client/index.ts]: JWTPayload type definition is Record<string, string>, so compiler does not check undefined. But added it to the function return types.

File: README.md
Comment (NikoHelle) [reply README.md]: Fixed.

Comment (akikoskinen) [reply README.md]: Please fix it in the original commit that introduced the typo.

Comment (NikoHelle) [reply README.md]: Ah, fixed there now.

File: src/client/index.ts
Comment (NikoHelle) [reply src/client/index.ts]: Merged some changes, so now exports not done.

Comment (akikoskinen) [reply src/client/index.ts]: `TokenPayload` is now used in commit 8f0941b5 but the type is only introduced later in commit f04063938. Should introduce the type at the same time (or earlier) as it's used.

Comment (NikoHelle) [reply src/client/index.ts]: Ah, yes thx, I didn't notice that. Went with just `JWTPayload` and removed the `TokenPayload` type and `KeycloakTokenPayload`.

**Review by akikoskinen — APPROVED** [review]

**Pull request merged**
